app/controllers/registers_controller.rb

```diff
@@ -121,56 +117,47 @@ def recover_entries_history(register_id, fields, params)
   def populate_register(register)
     register_data = @registers_client.get_register(register.name.parameterize, register.register_phase)
 
-    populate_user_data(register, register_data)
-    populate_system_data(register, register_data)
+    populate_data(register, register_data, 'user')
+    populate_data(register, register_data, 'system')
 
-    register.populated = true
     register.fields = register_data.get_field_definitions.map { |field| field.item.value['field'] }.join(',')
     register.save
   end
 
-  def populate_user_data(register, register_data)
+  def populate_data(register, register_data, entry_type)
     entries = []
     records = []
     count = 0
 
-    register_data.get_records_with_history.each do |record|
+    latest_entry = Entry.where(spina_register_id: register.id, entry_type: entry_type).order(:entry_number).reverse_order.limit(1).first
+    latest_entry_number = latest_entry.nil? ? 0 : latest_entry.entry_number
+
+    (entry_type == 'user' ? register_data.get_records_with_history(latest_entry_number) : register_data.get_metadata_records_with_history(latest_entry_number)).each do |record|
       previous_entry_number = nil
 
       record[:records].each_with_index do |value, idx|
         count += 1
-        new_entry = Entry.new(spina_register: register, timestamp: value.entry.timestamp, hash_value: value.item.hash, entry_number: value.entry.entry_number, previous_entry_number: previous_entry_number, entry_type: 'user', key: record[:key], data: value.item.value)
+        new_entry = Entry.new(spina_register: register, data: value.item.value, timestamp: value.entry.timestamp, hash_value: value.item.hash, entry_number: value.entry.entry_number, previous_entry_number: previous_entry_number, entry_type: entry_type, key: value.entry.key)
         entries.push(new_entry)
 
         if idx == record[:records].length - 1 # The last entry is the record
-          records.push(Record.new(spina_register: register, timestamp: value.entry.timestamp, hash_value: value.item.hash, entry_type: 'user', key: record[:key], data: value.item.value))
+          records.push(Record.new(spina_register: register, data: value.item.value, timestamp: value.entry.timestamp, hash_value: value.item.hash, entry_number: value.entry.entry_number, entry_type: entry_type, key: value.entry.key))
         end
 
         previous_entry_number = value.entry.entry_number
       end
 
       next unless (count % 1000).zero?
 
+      bulk_remove_existing_records(register, entry_type, records.map{ |record| record.key })
```

Comment:
Do we need to do the `bulk_remove_existing_records` and subsequent `save` as a Transaction to avoid a race condition whereby the records are deleted and the new ones not saved?

Reply:
Yes, this would be worth looking at.